contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,568 @@
+```
+kind: PersistentVolume
+apiVersion: v1
+metadata:
+name: local-raw-pv
+spec:
+  volumeMode: Block #proposed API change
+  capacity:
+    storage: 100Gi
+  local:
+    path: /dev/xvdc #device path
+  accessModes:
+    - ReadWriteOnce
+  persistentVolumeReclaimPolicy: Delete 
+```
+## Pod API Changes:
+To provide better specificy and ensure support of inline volumes, the following changes are proposed in the pod specification.
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-db
+spec:
+    containers
+    - name: mysql
+      image: mysql
+      volumeDevices: #proposed API change
+      - name: my-db-data
+        devicePath: /dev/xvda #proposed API change
+    volumes:
+    - name: my-db-data
+      persistentVolumeClaim:
+	claimName: raw-pvc
+```
+## Storage Class non-API Changes:
+For dynamic provisioning, it is assumed that values pass in the parameter section are opaque, thus the introduction of utilizing
+fsType in the StorageClass can be used by the provisioner to indicate how to create the volume. The proposal for this value is
+defined here:
+https://github.com/kubernetes/kubernetes/pull/45345 
+Therefore, a provisioner could potentially provision a block device and install the filesystem onto it by indicating the volumeMode
+as 'block' but the fsType as 'xfs'.
+This section is provided as a general guideline, but each provisioner may implement their parameters independent of what is defined
+here. It is our recommendation that the volumeMode in the PVC be the guidance for the provisioner and overrides the value given in the fstype. Therefore a provisioner should be able to ignore the fstype and provision a block device if that is what the user requested via the PVC and the provisioner can support this.
+
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: block-volume
+provisioner: kubernetes.io/local-block-glusterfs
+parameters:
+```
+The provisioner (if applicable) should validate the parameters and return and error if the combination specified is not supported.
+This also allows the use case for leveraging a Storage Class for utilizing pre-defined static volumes. By labeling the Persistent Volumes
+with the Storage Class, volumes can be grouped and used according to how they are defined in the class.
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: block-volume
+provisioner: no-provisioning 
+parameters:
+```
+
+# Pod Security Policy (PSP) Changes:
+Since the utilization of block devices can pose a risk to possible kernel manipulation by malicious users, it may be desirable for the administrator to restrict the usage entirely within a cluster. A similar convention is used with hostPath in that for some kubernetes 
+implmentations it is disabled by default.
+Thus, the PSP can define whether a validating pod can request the usage of such devices through the volumeDevices parameter.
```

Comment:
* whether or not a referenced PVC represents a block device or not is not known at pod creation time
* what happens if a block device PVC is referenced in a pod's `volumes` list, but not in a `volumeDevices` list (and is either not referenced at all or is only referenced in a `volumeMounts` list)? does that still "pose a risk to possible kernel manipulation"?
* what would you expect to be allowed, or to happen if a block device volume is defined in a pod's `volumes` list, but referenced via an item in the `volumeMounts` list, not the `volumeDevices` list?

Reply:
Validation needs to check that the PVC.volumeMode in volumeMounts is filesystem.  And that  PVC.volumeMode in volumeDevices is block.

Reply:
> Validation needs to check that the PVC.volumeMode in volumeMounts is filesystem. And that PVC.volumeMode in volumeDevices is block.

What validation are you referring to? The PVC is not available when PSP or the API server validate a pod.

Reply:
would it make more sense to also add another object similar to volumes: and call it devices: - therefore container.volumeMounts --> pod.volumes list  and container.volumeDevices --> pod.devices list ?

Reply:
Can we do multi object validation in an admission controller?

Reply:
> Can we do multi object validation in an admission controller?

if you do, you add ordering constraints on pod creation that are not present today, which is not good.

additionally, because a PVC can change after you create the pod, you have to handle the case where the PVC type doesn't match the way the pod references it anyway

finally, PSP is about authorization, and is tied to the permissions associated with the user creating the pod and the service account referenced by the pod. The things it allows have to be encapsulated in the pod, not referenced in another object that can later change.

Reply:
Most fields in the PVC should not be allowed to change once set; volumeMode should not be allowed to change.  If we can't validate at API creation time, then our only other option to validate the correctness of volumeMode is at runtime.

Reply:
delete/recreate-with-different parameters is indistinguishable from changing a value

Reply:
So then yes, we can only enforce at mount time.

Reply:
@liggitt @msau42 Okay, I removed this from the PV/PVC (though we can do that validation at mount) should the security constraint be on the volumeDevices as to whether or not they are allowed to use it. Just like hostPath or PersistentVolumeClaims. Therefore it could be validated in the PSP as that is part of the API changes for the pod sec. Does this make sense?